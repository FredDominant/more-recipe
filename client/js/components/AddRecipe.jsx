import React from 'react';
import { connect } from 'react-redux';

import uploadImage from '../utils/uploadImage';
import recipeValidator from '../validation/recipeValidator';
/**
 * @description componet holds form to add recipes
 *
 * @class AddRecipe
 * @extends {React.Component}
 */
class AddRecipe extends React.Component {
/**
 * Creates an instance of AddRecipe.
 * @param {any} props
 * @memberof AddRecipe
 */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      description: '',
      ingredients: '',
      directions: '',
      recipeImage: '',
      errors: {}
    };
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onUpload = this.onUpload.bind(this);
  }
  /**
   *
   * @returns {null} null
   * @param {any} event
   * @memberof AddRecipe
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
 *
 * @returns {null} null
 * @param {any} event
 * @memberof AddRecipe
 */
  onUpload(event) {
    const image = event.target.files[0];
    this.setState({ recipeImage: image });
  }
  /**
 *
 * @returns {null} null
 * @memberof AddRecipe
 */
  isValid() {
    const { errors, isValid } = recipeValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
   *
   * @returns {null} null
   * @memberof AddRecipe
   */
  addRecipe() {
    uploadImage(this.state.recipeImage)
      .then((response) => {
        // console.log('image url after axios is', response.data.secure_url);
        this.setState({ recipeImage: response.data.secure_url });
        // console.log('image url after updating state is', this.state.recipeImage);
        // dispatch this to store;
      })
      .catch((error) => {
        console.log(error);
      });
  }
  /**
 *
 * @returns {null} null
 * @param {any} event
 * @memberof AddRecipe
 */
  handleSubmit(event) {
    event.preventDefault();
    if (this.isValid()) {
      this.addRecipe();
    }
  }
  /**
   *
   * @returns {HTML} html
   * @memberof AddRecipe
   */
  render() {
    const errors = this.state.errors;
    return (
      <div>
        <div className="modal fade" id="addRecipe" tabIndex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
          <div className="modal-dialog" role="document">
            <div className="modal-content container">
              <div className="modal-header">
                <h5 className="modal-title"><span id="title"><h3 title="More Recipes and cooking tips">More Recipes</h3></span></h5>
                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <div className="modal-body">
                <div className="container error-body">
                  {errors.name && <div className="alert alert-dismissible alert-danger" role="alert">{errors.name}</div>}
                  {errors.description && <div className="alert alert-dismissible alert-danger" role="alert">{errors.description}</div>}
                  {errors.ingredients && <div className="alert alert-dismissible alert-danger" role="alert">{errors.ingredients}</div>}
                  {errors.directions && <div className="alert alert-dismissible alert-danger" role="alert">{errors.directions}</div>}
                </div>
                <form onSubmit={this.handleSubmit}>
                  <div className="form-group">
                    <label htmlFor="exampleInputEmail1">Recipe Name</label>
                    <input
                      type="text"
                      className="form-control"
                      name="name"
                      id="recipeName"
                      aria-describedby="small-recipe-name"
                      placeholder="Awesome Recipe!"
                      value={this.state.name}
                      onChange={this.onChange}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="recipe-description">Recipe Description:</label>
                    <input
                      type="text"
                      className="form-control"
                      name="description"
                      id="recipe-description"
                      placeholder="Awesome sauce made out of nothing!"
                      value={this.state.description}
                      onChange={this.onChange}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="cook-directions">Ingredients:</label>
                    <textarea
                      className="form-control"
                      id="cook-directions"
                      name="ingredients"
                      rows="3"
                      placeholder="Onions, tomatoes, grilled chicken, nutmeg"
                      value={this.state.ingredients}
                      onChange={this.onChange}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="cook-directions">Directions to cook:</label>
                    <textarea
                      className="form-control"
                      name="directions"
                      id="cook-directionss"
                      rows="3"
                      placeholder="dice tomatoes, tear chicken, boil onions, mix with nutmeg"
                      value={this.state.directions}
                      onChange={this.onChange}
                    />
                  </div>

                  <div className="form-group">
                    <label htmlFor="recipeImage">Recipe Image</label>
                    <input
                      type="file"
                      className="form-control-file"
                      name="file"
                      id="recipeImage"
                      onChange={this.onUpload}
                    />
                  </div>
                  <button className="btn btn-primary">Add Recipe</button>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({

});
export default connect(mapStateToProps)(AddRecipe);
