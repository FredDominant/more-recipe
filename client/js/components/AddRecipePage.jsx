import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import MDSpinner from 'react-md-spinner';
import toastr from 'toastr';

import uploadImage from '../utils/uploadImage';
import recipeValidator from '../validation/recipeValidator';
import addRecipe from '../actions/addRecipe';
/**
 *
 * @class AddRecipePage
 *
 * @extends {React.Component}
 */
export class AddRecipePage extends React.Component {
/**
 * @description Creates an instance of AddRecipePage.
 *
 * @param {any} props
 *
 * @memberof AddRecipePage
 */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      description: '',
      ingredients: '',
      directions: '',
      isUploading: false,
      recipeImage: '/images/noImage.png',
      errors: {},
      addRecipeError: '',
      uploadImageError: ''
    };
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onUpload = this.onUpload.bind(this);
    this.handleSelectImage = this.handleSelectImage.bind(this);
  }
  /**
 * @description component lifecycle method
 *
 * @returns {null} null
 *
 * @param {nextProp} nextProp object from store
 *
 * @memberof AddRecipe
 */
  componentWillReceiveProps(nextProp) {
    this.setState({ addRecipeError: nextProp.errorMessage });
  }

  /**
 *
 * @returns {null} null
 * @param {any} event
 * @memberof AddRecipe
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
*
* @returns {null} null

* @param {any} event

* @memberof AddRecipe
*/
  onUpload(event) {
    const reader = new FileReader();
    reader.onload = (e) => {
      this.setState({ recipeImage: e.target.result });
    };
    reader.readAsDataURL(event.target.files[0]);
  }
  /**
*
* @returns {null} null

* @memberof AddRecipe
*/
  isValid() {
    const { errors, isValid } = recipeValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
 *
 * @returns {null} null
 *
 * @memberof AddRecipe
 */
  addNewRecipe() {
    const { recipeImage, name, description, directions, ingredients } = this.state;
    if (recipeImage !== '/images/noImage.png') {
      this.setState({ isUploading: true });
      return uploadImage(recipeImage)
        .then((response) => {
          const recipeUrl = response.data.secure_url;
          console.log('recipe url is', recipeUrl);
          this.setState({ recipeImage: recipeUrl, isUploading: false });
          this.props.createRecipe(this.state);
        })
        .catch(() => {
          this.setState({ isUploading: false, uploadImageError: 'Could not upload image. Recipe not added' });
          toastr.options = {
            closeButton: true
          };
          toastr.error(this.state.uploadImageError);
        });
    }
    console.log('jumped to this place');
    this.props.createRecipe({ name, directions, ingredients, description });
  }
  /**
*
* @returns {null} null

* @param {any} event

* @memberof AddRecipe
*/
  handleSubmit(event) {
    event.preventDefault();
    if (this.isValid()) {
      this.addNewRecipe();
    }
  }
  /**
   *
   * @memberof AddRecipePage
   *
   * @returns {null} null
   */
  handleSelectImage() {
    $('input[type=file]').click();
  }
  /**
   *
   * @returns {null} null
   *
   * @memberof AddRecipePage
   */
  render() {
    const { errors, recipeImage, name, description, directions, ingredients } = this.state;
    return (
      <div className="container-fluid">
        <div id="add-recipe-form" className="container">
          <form onSubmit={this.handleSubmit}>
            <div className="row">
              <div className="col-sm-4">
                <div
                  className="recipe-image"
                  id="add-recipe-image-container"
                  onClick={this.handleSelectImage}
                  role="presentation"
                >
                  <img
                    className="img-thumbnail"
                    src={recipeImage}
                    alt=""
                    srcSet=""
                    id="add-recipe-image"
                  />
                </div>
                <br />
                <input
                  style={{ display: 'none' }}
                  type="file"
                  name="file"
                  id="file-upload"
                  onChange={this.onUpload}
                />
              </div>
              <div className="col-sm-8">
                <div className="form-group">
                  <label htmlFor="recipeName">Recipe Name</label>
                  <input
                    type="text"
                    id="recipeName"
                    className="form-control"
                    name="name"
                    value={name}
                    onChange={this.onChange}
                  />
                  {errors.name && <small className="form-text text-muted">
                    <span className="error-text"> {errors.name} </span>
                  </small>}
                </div>

                <div className="form-group">
                  <label htmlFor="recipeDescription">Recipe Description</label>
                  <input
                    type="text"
                    id="recipeDescription"
                    className="form-control"
                    name="description"
                    value={description}
                    onChange={this.onChange}
                  />
                  {errors.description && <small className="form-text text-muted">
                    <span className="error-text"> {errors.description} </span>
                  </small>}
                </div>

                <div className="form-group">
                  <label htmlFor="recipeDirections">Recipe Directions</label>
                  <textarea
                    type="text"
                    rows="5"
                    id="recipeDirections"
                    className="form-control"
                    name="directions"
                    value={directions}
                    onChange={this.onChange}
                  />
                  {errors.directions && <small className="form-text text-muted">
                    <span className="error-text"> {errors.directions} </span>
                  </small>}
                </div>

                <div className="form-group">
                  <label htmlFor="recipeingredients">Ingredients</label>
                  <textarea
                    type="text"
                    rows="5"
                    id="recipeIngredients"
                    className="form-control"
                    placeholder="Comma separated list of ingredients"
                    name="ingredients"
                    value={ingredients}
                    onChange={this.onChange}
                  />
                  {errors.ingredients && <small className="form-text text-muted">
                    <span className="error-text"> {errors.ingredients} </span>
                  </small>}
                </div>

                <div className="form-group">
                  <button className="btn btn-primary">
                    {
                      !this.state.isUploading &&
                      <h6 className="text-center"> Add Recipe </h6>
                    }
                    {
                      this.state.isUploading &&
                      <h6 className="text-center"> Uploading Image... </h6>
                    }
                    {
                      this.props.loading &&
                      <span> <MDSpinner /></span>
                    }
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  errorMessage: state.addRecipe.addRecipeErrorMessage,
  addRecipeSuccess: state.addRecipe.addRecipeSuccess,
  loading: state.isUploading
});

const mapDispatchToProps = dispatch => ({
  createRecipe: recipe => dispatch(addRecipe(recipe))
});

AddRecipePage.propTypes = {
  createRecipe: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired
};
AddRecipePage.defaultProps = {
  errorMessage: null,
  addRecipeSuccess: null
};
export default connect(mapStateToProps, mapDispatchToProps)(AddRecipePage);

