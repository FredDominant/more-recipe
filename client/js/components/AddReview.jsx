import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import addReview from '../actions/addReview';
/**
 *
 * @class AddReview
 *
 * @extends {React.Component}
 */
export class AddReview extends React.Component {
/**
 * @description Creates an instance of AddReview.
 *
 * @param {object} props
 *
 * @memberof AddReview
 */
  constructor(props) {
    super(props);
    this.state = {
      content: ''
    };
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  /**
   *
   * @param {object} event
   *
   * @memberof AddReview
   *
   * @returns {null} null
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
   * @returns {null} null
   *
   * @param {object} event
   *
   * @memberof AddReview
   */
  handleSubmit(event) {
    const { content } = this.state;
    const recipeId = this.props.recipeId;
    event.preventDefault();
    if (content.trim().length) {
      this.props.postReview(content, recipeId);
      this.setState({ content: '' });
    }
  }
  /**
   *
   *
   * @memberof AddReview
   *
   * @return {ReactElement} markup
   */
  render() {
    return (
      <div>
        <div id="review-form">
          <div>
            <div>
              <form onSubmit={this.handleSubmit}>
                <input
                  type="text"
                  className="form-control"
                  name="content"
                  placeholder="Add a review..."
                  onChange={this.onChange}
                  value={this.state.content}
                  required
                />
                <br />
                <button className="btn" id="add-review-button">
                  <span><i className="fas fa-comment-alt" /> </span>
                  Post Review
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
AddReview.propTypes = {
  postReview: PropTypes.func.isRequired,
  recipeId: PropTypes.number
};
AddReview.defaultProps = {
  recipeId: 0
};
const mapDispatchToProps = dispatch => ({
  postReview: (review, recipeId) => dispatch(addReview(review, recipeId))
});
export default connect(null, mapDispatchToProps)(AddReview);

