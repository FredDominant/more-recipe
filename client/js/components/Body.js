import React from 'react';
import ReactDOM from 'react-dom';
import { Switch, Route } from 'react-router-dom';

import { Home } from './Home';
import { Profile } from './Profile';
import { Recipe } from './Recipe'
import { UserHome } from './UserHome';

export class Body extends React.Component {
  render() {
    return (
			<Switch>
				<Route path='/' exact component={Home} />
				<Route path='/home' exact component={UserHome} />
				<Route path='/recipe' component={Recipe} />
				<Route path='/profile' component={Profile} />
			</Switch>
    );
  }
}
