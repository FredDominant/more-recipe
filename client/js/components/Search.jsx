import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import search from '../actions/search';
/**
 *
 * @export
 *
 * @class Search
 *
 * @extends {React.Component}
 */
class Search extends React.Component {
  /**
   * @description Creates an instance of Search.
   *
   * @param {object} props
   *
   * @memberof Search
   */
  constructor(props) {
    super(props);
    this.state = { searchParam: '' };
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onFocus = this.onFocus.bind(this);
  }
  /**
   *
   * @returns {null} null
   *
   * @param {object} event
   *
   * @memberof Search
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
    this.props.search(event.target.value);
  }
  /**
   * @memberof Search
   *
   * @returns {null} null
   */
  onFocus() {
    this.context.router.history.push('/search');
  }
  /**
   *
   * @returns {null} null
   *
   * @param {object} event
   *
   * @memberof Search
   */
  handleSubmit(event) {
    event.preventDefault();
    const { searchParam } = this.state;
    this.props.search(searchParam);
  }
  /**
   *
   * @returns {null} null
   *
   * @memberof Search
   */
  render() {
    return (
      <div className="container">
        <div className="search-body">
          <form onSubmit={this.handleSubmit}>
            <div className="input-group">
              <span id="search-intro"> <h5>Find a recipe: </h5>  </span>
              <input
                type="text"
                name="searchParam"
                id="search-box"
                className="form-control"
                placeholder="Search Recipe name or ingredient"
                aria-label="Search for..."
                value={this.state.search}
                onChange={this.onChange}
                onFocus={this.onFocus}
                autoFocus={this.props.focus}
              />
              <span className="input-group-btn">
                <button
                  className="btn btn-danger"
                  type="submit"
                >
                  <i className="fa fa-search" aria-hidden="true" /></button>
              </span>
            </div>
          </form>
        </div>
      </div>
    );
  }
}
Search.propTypes = {
  search: PropTypes.func.isRequired,
  focus: PropTypes.bool
};
Search.defaultProps = {
  focus: false
};

Search.contextTypes = {
  router: PropTypes.object.isRequired
};
const mapDispatchToProps = dispatch => ({
  search: searchParam => (dispatch(search(searchParam)))
});
export default connect(null, mapDispatchToProps)(Search);
