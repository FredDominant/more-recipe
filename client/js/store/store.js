const store = {
  loginForm: {
    email: '',
    password: ''
  },
  signunForm: {
    firstname: '',
    lastname: '',
    email: '',
    password: '',
    confirmPassword: ''
  }
};
export default store;
