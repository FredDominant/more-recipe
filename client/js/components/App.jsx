import React from 'react';

import Body from './Body';
import Errorboundary from '../components/Errorboundary';
import Navbar from './Navbar';

const App = () => (
  <div>
    <Navbar>
      <Errorboundary >
        <Body />
      </Errorboundary>
    </Navbar>
  </div>
);
export default App;
