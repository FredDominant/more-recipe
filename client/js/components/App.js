import React from 'react';
import ReactDOM from 'react-dom';

import { Search } from './Search';
import { Navigation } from './Navigation';
import { Body } from './Body';

 export class App extends React.Component {
  render() {
    return (
      <div>
					<Body />
      </div>
    );
  }
};
