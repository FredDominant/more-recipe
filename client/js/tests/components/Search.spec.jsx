import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import sinon from 'sinon';
import store from '../../store/store';

import { Search } from '../../components/Search';

const props = {
  search: jest.fn(),
};

/**
 * @description test for Search component
 *
 * @returns {null} null
 */
function setup() {
  const shallowWrapper = shallow(<Search {...props} />);
  const mountedWrapper = mount(
    <Provider store={store}>
      <BrowserRouter>
        <Search {...props} store={store} />
      </BrowserRouter>
    </Provider>
  );
  return {
    shallowWrapper,
    mountedWrapper
  };
}
describe('Search component', () => {
  it('should render correctly', () => {
    const { mountedWrapper } = setup();
    expect(mountedWrapper).toMatchSnapshot();
  });

  describe('onChange', () => {
    it('should set form value', () => {
      const event = {
        target: { name: 'searchParam', value: 'recipes' }
      };
      const { shallowWrapper } = setup();
      const searchField = shallowWrapper.find('#search-box');
      searchField.simulate('change', event);
      expect(shallowWrapper.instance().state.searchParam).toBe('recipes');
    });
  });

  describe('onFocus should', () => {
    it('redirect to another page', () => {
      sinon.spy(Search.prototype, 'onFocus');
      const { shallowWrapper } = setup();
      const searchField = shallowWrapper.find('#search-box');
      searchField.simulate('focus');
      expect(Search.prototype.onFocus.calledOnce).toEqual(true);
    });
  });

  describe('handleSubmit()', () => {
    it('to work if called', () => {
      const wrapper = shallow(<Search {...props} />);
      sinon.spy(Search.prototype, 'handleSubmit');
      const event = {
        preventDefault: jest.fn()
      };
      const form = wrapper.find('form');
      wrapper.setState({ serachParam: 'recipes' });
      form.simulate('submit', event);
      wrapper.instance().handleSubmit(event);
      const search = jest.spyOn(wrapper.instance().props, 'search');
      expect(search).toBeCalled();
      expect(Search.prototype.handleSubmit.calledOnce).toEqual(false);
    });
  });
});
