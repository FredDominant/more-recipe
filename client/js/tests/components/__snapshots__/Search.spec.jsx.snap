// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "blank",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Search
        focus={false}
        history={
          Object {
            "push": [Function],
          }
        }
        search={[MockFunction]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
      >
        <div
          className="container"
        >
          <div
            className="search-body"
          >
            <form
              onSubmit={[Function]}
            >
              <div
                className="input-group"
              >
                <span
                  id="search-intro"
                >
                   
                  <h5>
                    Find a recipe: 
                  </h5>
                    
                </span>
                <input
                  aria-label="Search for..."
                  autoFocus={false}
                  className="form-control"
                  id="search-box"
                  name="searchParam"
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Search Recipe name or ingredient"
                  type="text"
                />
                <span
                  className="input-group-btn"
                >
                  <button
                    className="btn btn-danger"
                    type="submit"
                  >
                    <i
                      aria-hidden="true"
                      className="fa fa-search"
                    />
                  </button>
                </span>
              </div>
            </form>
          </div>
        </div>
      </Search>
    </Router>
  </BrowserRouter>
</Provider>
`;
